### PR TITLE
Add analyzer that recommends using PeerAuthentication

### PR DESCRIPTION
**Pull request by selmanj**
The Policy/MeshPolicy resource is going away. This analyzer checks for their presence and suggests moving to the replacement PeerAuthentication resource. It does this by checking for the existence of the PeerAuthentication CRD.

Also added:

* Add support for CustomResourceDefinition in Galley
* Minor corrections to README.md for analyzers

### CODE REVIEW

**Comment by selmanj** [comment]
I'm not entirely happy with the contents of the actual message but I'm sending this out for review early to make sure I've understood things correctly re: the new PeerAuthentication api.

We'll also have to put together more substantial docs describing what to do when this message occurs (e.g. how to migrate). Is that already being tracked somewhere? If so I can move it into the docs for the analyzer code.

File: galley/pkg/config/analysis/analyzers/policy/policy.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	var hasPeerAuthenticationsCRD bool
+	c.ForEach(collections.K8SApiextensionsK8SIoV1Beta1Customresourcedefinitions.Name(),
+		func(r *resource.Instance) bool {
+			if r.Metadata.FullName.String() == "peerauthentications.security.istio.io" {
+				hasPeerAuthenticationsCRD = true
+				return false
+			}
+			return true
+		})
+	if !hasPeerAuthenticationsCRD {
```
Comment (diemtvu) [inline galley/pkg/config/analysis/analyzers/policy/policy.go]: Just to make sure I understand correctly, this check the CRD, not CRs, right. In other words, we do not show warning if user haven't upgrade to 1.5, hence have the CRD created.

Comment (selmanj) [reply galley/pkg/config/analysis/analyzers/policy/policy.go]: Exactly right. It's way easier to look for the CRD than it is to discover istio version (which isn't usually a well-defined concept).

Here we are just looking for the presence of the CRD, not any instances of CRs.

**Comment by selmanj** [comment]
Marking as do-not-merge. We need to figure out a better messaging strategy so this doesn't produce messages on freshly-installed clusters (as shown by the failing integration tests above).

@diemtvu do you know if fresh installs of istio 1.5 will use Policy/MeshPolicy resources, or will those entirely be replaced with PeerAuthentication resources? We should probably hold this PR until that's the case.

File: galley/pkg/config/analysis/msg/messages.yaml [hunk 2]
```diff
@@ -233,3 +233,16 @@ messages:
...
+    code: IST0120
+    level: Info
+    description: "The Policy resource is deprecated and will be removed in a future istio release. Migrate to the PeerAuthentication resource."
+    template: "The Policy resource is deprecated and will be removed in a future istio release. Migrate to the PeerAuthentication resource."
+
+  - name: "MeshPolicyResourceIsDeprecated"
+    code: IST0121
+    level: Info
+    description: "The MeshPolicy resource is deprecated and will be removed in a future istio release. Migrate to the PeerAuthentication resource."
+    template: "The MeshPolicy resource is deprecated and will be removed in a future istio release. Migrate to the PeerAuthentication resource."
```
Comment (esnible) [inline galley/pkg/config/analysis/msg/messages.yaml]: Capitalize `Istio` in all messages.

**Comment by jasonwzm** [comment]
I think we need a more unified solution to field/type deprecation in Istio CRDs as indicated in https://github.com/istio/istio/issues/19355. We have https://github.com/istio/istio/blob/master/galley/pkg/config/analysis/analyzers/deprecation/deprecation.go as a field deprecation analyzer for now, maybe we can extend that analyzer to include the deprecation cases?

**Comment by diemtvu** [comment]
> sh installs of istio 1.5 will use Policy/MeshPolicy resources, or will those entirely be replaced with PeerAuthentication resources? We should probably

I am working on the PR for installation. It should not install the mesh policy config. In permissive mode, it should not add any policy (resource) neither.

**Comment by selmanj** [comment]
@diemtvu Can you link to the PR? I'd consider this one blocked until that's in place.

File: galley/pkg/config/analysis/msg/messages.yaml (hunk 2 above)
Comment (selmanj) [reply galley/pkg/config/analysis/msg/messages.yaml]: Done, thanks!

**Comment by jacob-delgado** [comment]
> I think we need a more unified solution to field/type deprecation in Istio CRDs as indicated in #19355. We have [/galley/pkg/config/analysis/analyzers/deprecation/deprecation.go@`master`](https://github.com/istio/istio/blob/master/galley/pkg/config/analysis/analyzers/deprecation/deprecation.go) [](/istio/istio/blob/HEAD@{2020-02-10T18:57:18Z}/galley/pkg/config/analysis/analyzers/deprecation/deprecation.go) as a field deprecation analyzer for now, maybe we can extend that analyzer to include the deprecation cases?

There is also [this message](https://github.com/istio/istio/blob/b80eec9debb233b62cef2cd8b13ef14972587558/galley/pkg/config/analysis/analyzers/deprecation/deprecation.go#L122) for Policy using JWT authentication deprecation.

v1beta1/PeerAuthentication is for mTLS policy and does not handle JWT authentication.

By suggesting that we always suggest using v1beta1/PeerAuthentication it may have mislead users. Although checking for the presence of the new CRDs is a very good step forward.

**Comment by selmanj** [comment]
@jacob-delgado good point! I think once we have proper documentation written on exactly how to migrate we can move forward.

@jasonwzm and I chatted a bit about how to tackle this more generally, especially around the challenges of pointing users to relevant documentation.

**Comment by diemtvu** [comment]
> @diemtvu Can you link to the PR? I'd consider this one blocked until that's in place.

It's https://github.com/istio/istio/pull/21017. However, the env WG suggest to keep it as is to be safe :(. That means Upgrade/new users will still see the MeshConfig in old API. Though our docs and tools should still steer them away. Does this make sense?

**Comment by selmanj** [comment]
Spoke to @diemtvu, it sounds like the default install will not come with any alpha security CRs defined. Seems like we should be go to merge this then.

**Comment by yangminzhu** [comment]
This is a nice change, it would be better to give the same deprecation warning for  `ClusterRbacConfig`, `ServiceRole` and `ServiceRoleBinding` for the old RBAC that is deprecated by the new `AuthorizationPolicy` since 1.4. (The old RBAC will be removed completely in 1.6).

@selmanj Do you think you will have some cycles to also add this for authorization? Thank you.

**Comment by yangminzhu** [comment]
Also should this be cherry-picked to release-1.5?

**Comment by selmanj** [comment]
@yangminzhu I unfortunately won't be able to, but I think @jasonwzm expressed interest in solving this more generally across different resource types.

**Pull request merged**